cpp/src/parquet/stream_writer.h

```diff
@@ -0,0 +1,181 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_STREAM_WRITER_H
+#define PARQUET_STREAM_WRITER_H
+
+#include <array>
+#include <chrono>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/util/string_view.h"
+#include "parquet/api/writer.h"
+
+namespace parquet {
+
+// A class for writing Parquet files using an output stream type API.
+//
+// The values given must be of the correct type i.e. the type must
+// match the file schema exactly otherwise a ParquetException will be
+// thrown.
+//
+// The user must explicitly indicate the end of the row using the
+// EndRow() function or EndRow output manipulator.
+//
+// A maximum row group size can be configured, the default size is
+// 512MB.  Alternatively the row group size can be set to zero and the
+// user can create new row groups by calling the EndRowGroup()
+// function or using the EndRowGroup output manipulator.
+//
+// Currently there is no support for optional or repeated fields.
+//
+class PARQUET_EXPORT StreamWriter {
+ public:
+  StreamWriter() = default;
+
+  explicit StreamWriter(std::unique_ptr<ParquetFileWriter>& File_Writer);
+
+  ~StreamWriter();
+
+  static void SetDefaultMaxRowGroupSize(const int64_t max_size);
```

Comment:
we generally don't use const on primitive types. (https://google.github.io/styleguide/cppguide.html#Use_of_const)

Reply:
Although a const parameter passed by value has no effect on the caller, it does guarantee the implementationi cannot modify the parameter given to it, which is nice to know and avoids errors.

Reply:
*shrug* style ... same comment as above, we can let parquet committer be a tie breaker.  FWIW, you can keep const out of the declarations and only put them on the definition where they add this value (https://abseil.io/tips/109)

Reply:
I would prefer to not use const for primitive types for consistency with the rest of the codebase. It's a detail that need not be present in the public API